File: backend/app/utils/logger.py
"""
Centralized logging service for the application.
Provides structured logging with context and different log levels.
"""
import logging
import json
from datetime import datetime
from typing import Any, Dict, Optional
import sys


class ColoredFormatter(logging.Formatter):
    """Custom formatter with colors for console output"""

    COLORS = {
        'DEBUG': '\033[36m',     # Cyan
        'INFO': '\033[32m',      # Green
        'WARNING': '\033[33m',   # Yellow
        'ERROR': '\033[31m',     # Red
        'CRITICAL': '\033[35m',  # Magenta
    }
    RESET = '\033[0m'

    def format(self, record):
        levelname = record.levelname
        if levelname in self.COLORS:
            record.levelname = f"{self.COLORS[levelname]}{levelname}{self.RESET}"
        return super().format(record)


class StructuredLogger:
    """Structured logger with context support"""

    def __init__(self, name: str):
        self.logger = logging.getLogger(name)
        self.logger.setLevel(logging.DEBUG)

        # Console handler with colors
        if not self.logger.handlers:
            console_handler = logging.StreamHandler(sys.stdout)
            console_handler.setLevel(logging.DEBUG)
            formatter = ColoredFormatter(
                '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                datefmt='%Y-%m-%d %H:%M:%S'
            )
            console_handler.setFormatter(formatter)
            self.logger.addHandler(console_handler)

    def _format_message(self, message: str, context: Optional[Dict[str, Any]] = None) -> str:
        """Format message with optional context"""
        if context:
            context_str = json.dumps(context, default=str)
            return f"{message} | Context: {context_str}"
        return message

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log debug message"""
        self.logger.debug(self._format_message(message, context))

    def info(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log info message"""
        self.logger.info(self._format_message(message, context))

    def warning(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log warning message"""
        self.logger.warning(self._format_message(message, context))

    def error(self, message: str, context: Optional[Dict[str, Any]] = None, exc_info: bool = False):
        """Log error message"""
        self.logger.error(self._format_message(message, context), exc_info=exc_info)

    def critical(self, message: str, context: Optional[Dict[str, Any]] = None, exc_info: bool = False):
        """Log critical message"""
        self.logger.critical(self._format_message(message, context), exc_info=exc_info)


# Pre-configured loggers for different parts of the application
api_logger = StructuredLogger('api')
db_logger = StructuredLogger('database')
auth_logger = StructuredLogger('auth')
service_logger = StructuredLogger('service')
websocket_logger = StructuredLogger('websocket')

File: src/utils/logger.ts
/**
 * Centralized logging utility for the frontend application
 * Provides structured logging with different log levels and contexts
 */

export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR',
}

interface LogContext {
  [key: string]: any;
}

class Logger {
  private name: string;
  private colors = {
    DEBUG: '#6B7280', // Gray
    INFO: '#3B82F6',  // Blue
    WARN: '#F59E0B',  // Orange
    ERROR: '#EF4444', // Red
  };

  constructor(name: string) {
    this.name = name;
  }

  private formatMessage(level: LogLevel, message: string, context?: LogContext): void {
    const timestamp = new Date().toISOString();
    const color = this.colors[level];

    const logArgs = [
      `%c[${timestamp}] ${level} [${this.name}]%c ${message}`,
      `color: ${color}; font-weight: bold`,
      'color: inherit',
    ];

    if (context) {
      logArgs.push('\nContext:', context);
    }

    switch (level) {
      case LogLevel.DEBUG:
        console.debug(...logArgs);
        break;
      case LogLevel.INFO:
        console.log(...logArgs);
        break;
      case LogLevel.WARN:
        console.warn(...logArgs);
        break;
      case LogLevel.ERROR:
        console.error(...logArgs);
        break;
    }
  }

  debug(message: string, context?: LogContext): void {
    this.formatMessage(LogLevel.DEBUG, message, context);
  }

  info(message: string, context?: LogContext): void {
    this.formatMessage(LogLevel.INFO, message, context);
  }

  warn(message: string, context?: LogContext): void {
    this.formatMessage(LogLevel.WARN, message, context);
  }

  error(message: string, error?: Error | any, context?: LogContext): void {
    const errorContext = {
      ...context,
      error: error instanceof Error ? {
        message: error.message,
        stack: error.stack,
        name: error.name,
      } : error,
    };
    this.formatMessage(LogLevel.ERROR, message, errorContext);
  }

  // Performance tracking
  time(label: string): void {
    console.time(`[${this.name}] ${label}`);
  }

  timeEnd(label: string): void {
    console.timeEnd(`[${this.name}] ${label}`);
  }

  // Group related logs
  group(label: string): void {
    console.group(`[${this.name}] ${label}`);
  }

  groupEnd(): void {
    console.groupEnd();
  }
}

// Pre-configured loggers for different parts of the application
export const apiLogger = new Logger('API');
export const authLogger = new Logger('Auth');
export const uiLogger = new Logger('UI');
export const wsLogger = new Logger('WebSocket');
export const rtcLogger = new Logger('WebRTC');

// Factory function to create custom loggers
export function createLogger(name: string): Logger {
  return new Logger(name);
}

export default Logger;
